File: src/main/java/de/cplaiz/activecraftcore/manager/MessageManager.java
package de.cplaiz.activecraftcore.manager;

import de.cplaiz.activecraftcore.ActiveCraftCore;
import de.cplaiz.activecraftcore.events.MsgEvent;
import de.cplaiz.activecraftcore.exceptions.InvalidPlayerException;
import de.cplaiz.activecraftcore.messagesv2.MessageFormatter;
import de.cplaiz.activecraftcore.messagesv2.MessageSupplier;
import de.cplaiz.activecraftcore.messagesv2.PlayerMessageFormatter;
import de.cplaiz.activecraftcore.playermanagement.OfflinePlayerActionScheduler;
import de.cplaiz.activecraftcore.playermanagement.Profilev2;
import lombok.Getter;
import org.bukkit.Bukkit;
import org.bukkit.Sound;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;

import java.util.HashMap;

public class MessageManager {

    @Getter
    private static final HashMap<CommandSender, Profilev2> msgPlayerStoring = new HashMap<>();

    public static void sendDM(Profilev2 receiver, CommandSender sender, String message) {
        MessageSupplier acCoreMessageSupplier = receiver.getMessageSupplier(ActiveCraftCore.getInstance());
        MessageFormatter msgFormatter = new PlayerMessageFormatter(ActiveCraftCore.getInstance().getActiveCraftMessagev2())
                .setTarget(receiver).setSender(sender);
        MsgEvent event = new MsgEvent(sender, receiver, message);
        Bukkit.getPluginManager().callEvent(event);
        if (event.isCancelled()) return;
        if (sender instanceof Player) {
            OfflinePlayerActionScheduler.schedule(receiver, (profile) -> {
                msgFormatter.addFormatterPattern("message", event.getMessage());
                Player target = receiver.getPlayer();
                target.sendMessage(acCoreMessageSupplier.getFormatted("command.msg.format-from", msgFormatter));
                target.playSound(target.getLocation(), Sound.ENTITY_ARROW_HIT_PLAYER, 1f, 1f);
                msgPlayerStoring.put(target, Profilev2.of(sender.getName()));
                Bukkit.getOnlinePlayers().stream()
                        .filter(player -> player.hasPermission("activecraft.msg.spy") && receiver.canReceiveSocialspy() && (player != sender && player != target))
                        .forEach(player -> player.sendMessage(acCoreMessageSupplier.getFormatted("command.msg.socialspy-format", msgFormatter)));
                if (ActiveCraftCore.getInstance().getMainConfig().isSocialSpyToConsole())
                    Bukkit.getConsoleSender().sendMessage(acCoreMessageSupplier.getFormatted("command.msg.socialspy-format", msgFormatter));
                return null;
            });
        } else {
            OfflinePlayerActionScheduler.schedule(receiver, (profile) -> {
                Player target = receiver.getPlayer();
                msgFormatter.addFormatterPattern("message", event.getMessage());
                target.sendMessage(acCoreMessageSupplier.getFormatted("command.msg.from-console-format", msgFormatter));
                target.playSound(target.getLocation(), Sound.ENTITY_ARROW_HIT_PLAYER, 1f, 1f);
                return null;
            });
        }
    }

    public static void openConversation(CommandSender sender, Profilev2 target) {
        msgPlayerStoring.put(sender, target);
    }

    public static void msgToActiveConversation(CommandSender sender, String message) throws InvalidPlayerException {
        if (msgPlayerStoring.get(sender) == null)
            throw new InvalidPlayerException("No open conversation");
        sendDM(msgPlayerStoring.get(sender), sender, message);
    }
}

File: src/main/java/org/activecraft/activecraftcore/manager/MessageManager.kt
package org.activecraft.activecraftcore.manager

import org.activecraft.activecraftcore.ActiveCraftCore
import org.activecraft.activecraftcore.events.MsgEvent
import org.activecraft.activecraftcore.exceptions.InvalidPlayerException
import org.activecraft.activecraftcore.messages.MessageFormatter
import org.activecraft.activecraftcore.messages.MessageSupplier
import org.activecraft.activecraftcore.messages.PlayerMessageFormatter
import org.activecraft.activecraftcore.playermanagement.OfflinePlayerActionScheduler.schedule
import org.activecraft.activecraftcore.playermanagement.Profile
import org.bukkit.Bukkit
import org.bukkit.Sound
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

object MessageManager {
    private val msgPlayerStoring = HashMap<CommandSender, Profile>()
    fun sendDM(receiver: Profile, sender: CommandSender, message: String) {
        val acCoreMessageSupplier: MessageSupplier = receiver.getMessageSupplier(ActiveCraftCore.INSTANCE)!!
        val msgFormatter: MessageFormatter =
            PlayerMessageFormatter(ActiveCraftCore.INSTANCE.activeCraftMessage!!)
                .setTarget(receiver).setSender(sender)
        val event = MsgEvent(sender, receiver, message)
        Bukkit.getPluginManager().callEvent(event)
        if (event.cancelled) return
        if (sender is Player) {
            schedule(receiver) {
                msgFormatter.addFormatterPattern("message", event.message)
                val target = receiver.player
                target!!.sendMessage(acCoreMessageSupplier.getFormatted("command.msg.format-from", msgFormatter))
                target.playSound(target.location, Sound.ENTITY_ARROW_HIT_PLAYER, 1f, 1f)
                msgPlayerStoring[target] = Profile.of(sender)
                Bukkit.getOnlinePlayers()
                    .filter { player: Player -> player.hasPermission("activecraft.msg.spy") && receiver.receiveSocialspy && player !== sender && player !== target }
                    .forEach { player: Player ->
                        player.sendMessage(
                            acCoreMessageSupplier.getFormatted(
                                "command.msg.socialspy-format",
                                msgFormatter
                            )
                        )
                    }
                if (ActiveCraftCore.INSTANCE.mainConfig.socialSpyToConsole) Bukkit.getConsoleSender()
                    .sendMessage(acCoreMessageSupplier.getFormatted("command.msg.socialspy-format", msgFormatter))
            }
        } else {
            schedule(receiver) {
                val target = receiver.player
                msgFormatter.addFormatterPattern("message", event.message)
                target!!.sendMessage(
                    acCoreMessageSupplier.getFormatted(
                        "command.msg.from-console-format",
                        msgFormatter
                    )
                )
                target.playSound(target.location, Sound.ENTITY_ARROW_HIT_PLAYER, 1f, 1f)
            }
        }
    }

    fun openConversation(sender: CommandSender, target: Profile) {
        msgPlayerStoring[sender] = target
    }

    @Throws(InvalidPlayerException::class)
    fun msgToActiveConversation(sender: CommandSender, message: String) {
        sendDM(msgPlayerStoring[sender] ?: throw InvalidPlayerException("No open conversation"), sender, message) // TODO: NoOpenConversationException
    }

    fun getMsgPartner(sender: CommandSender) = msgPlayerStoring[sender]
}